README.pod
=pod

=head1 NAME

DSL::Tiny::Role - A simple yet powerful DSL builder.

=head1 VERSION

version 0.001

=head1 SYNOPSIS

    # In e.g. MooseDSL.pm, describe a simple DSL.
    package MooseDSL;

    use Moose;  # or use Moo;

    with qw(DSL::Tiny::Role);

    sub build_dsl_keywords {
        return [
            # keywords will be run through curry_method
            qw(argulator return_self clear_call_log),
        ];
    }

    has call_log => (
        clearer => 'clear_call_log',
        default => sub { [] },
        is      => 'rw',
        lazy    => 1
    );

    sub argulator {
        my $self = shift;
        push @{ $self->call_log }, join "::", @_;
    }

    sub return_self { return $_[0] }

    1;

    ################################################################

    # and then in another file you can use that DSL

    use Test::More;
    use Test::Deep;

    use MooseDSL qw( -install_dsl );

    # peek under the covers, get the instance
    my $dsl = return_self;
    isa_ok( $dsl, 'MooseDSL' );

    # test argument handling, single scalar
    argulator("a scalar");
    cmp_deeply( $dsl->call_log, ['a scalar'], 'scalar arg works' );
    clear_call_log;

    # test argument handling, list of args
    argulator(qw(a list of things));
    cmp_deeply( $dsl->call_log, ['a::list::of::things'], 'list arg works' );
    clear_call_log;

    done_testing;

=head1 DESCRIPTION

I<This is an initial release.  It's all subject to rethinking.  Comments
welcome.>

    every time a language advertises "we make writing dsls easy!" i
    read "i'm going to have to learn a new language for every project"

    Jesse Luehrs (@doyster) 3/8/13, 12:11 PM

Domain-specific languages (DSL's) aid in the efficient expression of
configurations, problems and solutions within a particular domain.  While some
DSL's are built from the ground up with custom lexers, parsers,
etc... (e.g. the Unix build tool "make"), other "internal DSL's" (L<Werner
Schuster|http://www.infoq.com/news/2007/06/dsl-or-not>) are distilled from
existing languages and "speak the language of their domain with an accent"
(L<Piers Cawley|http://www.bofh.org.uk/2007/05/19/domain-agnostic-languages>).

A variety of Perl tools and libraries sport domain specific langagues,
e.g. L<Dancer>, L<Module-CPANfile> and L<Module-Install> and the number of
re-implementations of the underlying plumbing is almost exactly equal to the
number of such modules.  These implementations usually devolve into dirty
tricks (e.g. explicit package stash manipulations) and re-invention of several
wheels.

L<DSL::Tiny> packages the common functionality required to implement an
internal DSL, building on powerful foundations (L<Sub::Exporter>) and effective
techniques (L<Moose> and L<Moo> roles) to allow developers to focus on their
domain-specific issues.  It builds on a flexible mechanism for exporting a set
of subroutines into a package; provides a consistent framework for subroutine
currying; and automates the construction of instances, their association with
DSL fragments and the evaluation of those fragments.

In other words, when I needed to build an internal DSL for a project, I was
shocked at how often the basic brushstrokes had been repeated and how often
these implementations dug down and peeked underneath Perl's stashes.  These
modules are my attempt to provide a reusable solution to the problem via
existing high-leverage tools.

=head1 ATTRIBUTES

=head2 dsl_keywords

Returns an arrayref of dsl keyword info.

It is lazy.  Classes which consume the role are required to supply a builder
named C<_build_dsl_keywords>.

In its canonical form the contents of the array reference are a series of array
references containing keyword_name => { option_hash } pairs, e.g.

  [ [ keyword1 => { as => &generator('method1') } ],
    [ keyword2 => { before => &generator ]
  ]

Generators are as described in the L<Sub::Exporter> documentation.

However, as the contents of this array reference are processed with
Data::OptList there is a great deal of flexibility, e.g.

  [ qw( m1 m2 ), k4 => { as => &generator('some_method' } ]

is equivalent to:

  [ m1 => undef, m2 => undef, k4 => { as => generator('some_method') } ]

Options are optional.  In particular, if no C<as> generator is provided then
the keyword name is presumed to also be the name of a method in the class and
C<Sub::Exporter::Utils::curry_method> will be applied to that method to
generate the coderef for that keyword.  The makes the above equivalent to:

  [ m1 => { as => generator('m1') }, m2 => { as => generator('m2') },
    k4 => { as => generator('some_method') }
  ]

In its simplest form, the keyword arrayref contains a list of method names
relative to class which consumes this role.

  [ qw( m1 m2 ) ]

Supported options include:

=over 4

=item as

=item before

=item after

=back

=head1 METHODS

=head2 import

An import routine generated by Sub::Exporter.

When invoked as a class method (usually via C<use>) with a C<-install_dsl>
argument it will construct a new instance then generate and install a set of
subroutines using the information provided in the instance's C<dsl_keywords>
attribute.

TODO.

=head2 install_dsl

A synonym for the Sub::Exporter generated import method.  Sounds better when
one uses it to install into an instance.

=head2 _dsl_build

Private-ish.  Do you really want to be here?

C<_dsl_build> build's up the set of keywords that L<Sub::Exporter> will
install.

It returns a hashref whose keys are names of keywords and whose values are
coderefs implementing the respective behavior.

It can be invoked on a class (a.k.a. as a class method), usually by C<use>.  If
so, a new instance of the class will be constructed and the various keywords
are curried with respect to that instance.

It can be invoked on a class instance, e.g. via an explicit invocation of
L<import> on an instance.  If so, then that instance is used when constructing
the keywords.

=head2 _compile_keyword

Private, go away.

Generate a sub that implements the keyword, taking care of before's and afters.

=head1 REQUIRES

=head2 build_dsl_keywords

A subroutine, used as the Moo{,se} builder for the L</dsl_keywords> attribute.
It returns an array reference containing information about the methods and
subroutines that implement the keywords in the DSL.

=head1 AUTHOR

George Hartzell <hartzell@alerce.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by George Hartzell.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.



Changes
Revision history for {{$dist->name}}

{{$NEXT}}

0.001     2013-05-28 10:23:26 America/Los_Angeles
        - Initial release.


t/lib/MyDSL.pm
#!perl

package MyDSL;

use Moo;

with qw(DSL::Tiny::Role);

use MyDSL::Util qw(curry_with_tracer_call naked_generator);
use MyHelper;
use Sub::Exporter::Util qw(curry_chain curry_method);

sub build_dsl_keywords {
    return [
        # simple keyword -> curry_method examples
        qw(call_log_as_array clear_call_log),

        # simple keyword, written differently
        'my_context',

        # test explicit call to curry_method w/out renaming.
        main => { as => curry_method, },

        # use different keyword/method names and use curry_method
        break_encapsulation => { as => curry_method('return_self'), },

        # chain through has_a relationship
        test_curry_chain => { as => curry_chain( a_helper => 'beep' ), },

        # chain through has_a relationship, with an explicit warning to
        # helper's method.
        # ends up as $dsl->a_helper->beep( warning => 'Andele!') ]
        test_curry_chain_with_arg => {
            as => curry_chain( a_helper => beep => [ warning => 'Andele!' ] ),
        },

        # a single before generator
        test_simple_before => {
            as     => curry_method('main'),
            before => curry_method('before_1'),
        },

        # several before generators
        test_multi_before => {
            as     => curry_method('main'),
            before => [ curry_method('before_1'), curry_method('before_2'), ],
        },

        # a single after generator
        test_simple_after => {
            after => curry_method('after_1'),
            as    => curry_method('main'),
        },

        # several after generators
        test_multi_after => {
            after => [ curry_method('after_1'), curry_method('after_2'), ],
            as    => curry_method('main'),
        },

        # a bit of everything
        test_complex => {
            after  => [ curry_method('after_1'),  curry_method('after_2'), ],
            as     => curry_method('main'),
            before => [ curry_method('before_1'), curry_method('before_2'), ],
        },

        # a method that stuffs its args into the call log
        qw(argulator),

        qw(clear_trace_log),
        test_alternate_currier => { as => curry_with_tracer_call('main'), },

        naked => { as => naked_generator },
        bare  => { as => naked_generator('naked') },

    ];
}

has a_helper => ( is => 'rw', default => sub { return MyHelper->new() } );

has call_log => (
    clearer => 'clear_call_log',
    default => sub { [] },
    is      => 'rw',
    lazy    => 1
);

has trace_log => (
    clearer => 'clear_trace_log',
    default => sub { [] },
    is      => 'rw',
    lazy    => 1
);

sub after_1 { push @{ $_[0]->call_log }, 'after_1' }
sub after_2 { push @{ $_[0]->call_log }, 'after_2' }

sub argulator {
    my $self = shift;
    push @{ $self->call_log }, join "::", @_;
}

sub before_1 { push @{ $_[0]->call_log }, 'before_1' }
sub before_2 { push @{ $_[0]->call_log }, 'before_2' }

sub call_log_as_array { @{ $_[0]->call_log } }

sub main { push @{ $_[0]->call_log }, 'main' }

sub naked { return "buck" }

sub my_context {
    my $self = shift;
    if ( defined wantarray ) {
        if (wantarray) {
            push @{ $self->call_log }, 'array_context';
        }
        else {
            push @{ $self->call_log }, 'scalar_context';
        }
    }
    else {
        push @{ $self->call_log }, 'void_context';
    }
}

sub return_self { return $_[0] }

1;


t/lib/MyHelper.pm

package MyHelper;
use Moo;

sub beep { return "beep beep" }

1;


t/lib/MyEvalDSL.pm
#!perl

package MyEvalDSL;

use Moo;

with qw(DSL::Tiny::Role DSL::Tiny::InstanceEval);

use MyDelegate;
use Sub::Exporter::Util qw(curry_chain curry_method);

sub build_dsl_keywords {
    return [
        qw(ape incr site value),
        baz  => { as => curry_method('bar'), },
        beep => {
            as    => curry_chain( delegate => 'beep' ),
            after => curry_method('do_it_after'),
        },
        blap => {
            before => curry_method('do_it_before'),
            as     => curry_method('site'),
        },
    ];
}

has counter => ( is => 'rw', default => sub {0}, );

sub do_it_before {
    my $self = shift;
    print "before counter = " . $self->counter . "\n";
}

sub do_it_after {
    my $self = shift;
    print "after counter = " . $self->counter . "\n";
}

sub ape { return "DSL sez: ape" }

sub bar { return $_[0] }

sub delegate { return MyDelegate->new() }

sub incr { $_[0]->counter( $_[0]->counter() + 1 ) }

sub site { return "DSL sez: site" }

sub value { $_[0]->counter }

1;


lib/DSL/Tiny/Class.pm
package DSL::Tiny::Class;

use Moo;

use Sub::Exporter -setup => { groups => { install_dsl => \&dsl_build } };

use Data::OptList;

sub dsl_build {
    my ( $invocant, $group, $arg ) = @_;

    my $instance = ref $invocant ? $invocant : $invocant->new();

    my $keywords = Data::OptList::mkopt_hash( $instance->dsl_keywords,
        { moniker => 'keyword list' }, ['HASH'], );

    my %dsl = map { $_ => $instance->compile_keyword( $_, $keywords->{$_} ) }
        keys $keywords;

    return \%dsl;
}

sub dsl_keywords {
    return [ qw(ape incr site value), baz => { method => 'bar' } ];
}

sub compile_keyword {
    my ( $self, $keyword, $args ) = @_;

    my $method = $args->{method} || $keyword;

    return sub { $self->$method(@_) };
}

has counter => ( is => 'rw', default => sub {0}, );

sub ape { return "DSL sez: ape" }

sub bar { return $_[0] }

sub incr { $_[0]->counter( $_[0]->counter() + 1 ) }

sub site { return "DSL sez: site" }

sub value { $_[0]->counter }

1;


lib/DSL/Tiny/InstanceEval.pm
## no critic (RequireUseStrict RequireUseWarnings)
package DSL::Tiny::InstanceEval;
## critic
# ABSTRACT: Add DSL features to your class.

=head1 SYNOPSIS

    use Test::More;
    use Test::Deep;

    # put together class with a simple dsl
    {
      package MyClassWithDSL;
      use Moo;                      # or Moose
      with qw(DSL::Tiny::Role DSL::Tiny::InstanceEval);

      sub _build_dsl_keywords { [ qw(add_values) ] };

      has values => (is => 'ro',
                     default => sub { [] },
                    );

      sub add_values {
          my $self = shift;
          push @{$self->values}, @_;
      }
    }

    # make a new instance
    my $dsl = MyClassWithDSL->new();

    my $code = <<EOC;
    add_values(qw(2 1));
    add_values(qw(3));
    EOC

    my $return_value = $dsl->instance_eval($code);
    cmp_deeply($dsl->values, bag(qw(1 2 3)), "Values were added");

    done_testing;

=head1 DESCRIPTION

I<This is an initial release.  It's all subject to rethinking.  Comments
welcome.>

This package provides a simple interface, L</instance_eval>, for evaluating
snippets of a DSL (implemented with L<DSL::Tiny::Role>) with respect to a
particular instance of a class that consumes the role.

=cut

use Moo::Role;

use MooX::Types::MooseLike::Base qw(CodeRef Str);

=attr _anon_pkg_name

Private attribute, used to set up a package to stash private stuff.

=cut

has _anon_pkg_name => (
    is      => 'ro',
    isa     => Str,
    lazy    => 1,
    builder => '_build__anon_pkg_name',
);

{
    # no one can see me if I have my curly braces over my eyes....
    my $ANON_SERIAL = 0;

    # close over $ANON_SERIAL
    sub _build__anon_pkg_name { ## no critic(ProhibitUnusedPrivateSubroutines)
        return __PACKAGE__ . "::ANON_" . ++$ANON_SERIAL;
    }
}

=attr _instance_evalator

PRIVATE

There is no 'u' in _instance_evalator.  That means there should be no
you in there either....

Returns a coderef that is used by the instance_eval() method.

=cut

has _instance_evalator => (
    is       => 'ro',
    isa      => CodeRef,
    lazy     => 1,
    builder  => '_build__instance_evalator',
    init_arg => undef,
);

##
## - set up an environment (anonymous package) in which to execute code that is
##   being instance_eval'ed,
## - push curried closures into the package for each of the closures,
## - and build a coderef that switches to that package, does the eval,
##   dies if the eval had trouble and otherwise returns the eval's return value.
##
sub _build__instance_evalator { ## no critic(ProhibitUnusedPrivateSubroutines)
    my $self = shift;

    # make up a fairly unique package
    my $pkg_name = $self->_anon_pkg_name();

    # stuff the DSL into the fairly unique package
    $self->install_dsl( { into => $pkg_name }, qw(-install_dsl) );

    # stuff an evalator routine into the same package,
    # closed over $pkg_name
    # evals a string, dies if there was trouble, returns result otherwise.

    # return a coderef to the evalator routine that
    # we pushed into the package.
    return sub {
        my $code = 'package ' . $pkg_name . '; ' . shift;
        my $result = eval $code;    ## no critic (ProhibitStringyEval)
        die $@ if $@;               ## no critic (RequireCarping)
        return $result;
    };
}

=method instance_eval

Something kind-a-similar to Ruby's instance_eval.  Takes a string and evaluates
it using eval(), The evaluation happens in a package that has been populated
with a set of functions that map to methods in this class with the instance
curried out.

See the synopsis for an example.

=cut

sub instance_eval {    ## no critic(RequireArgUnpacking)
    my $self = shift;

    return $self->_instance_evalator()->(@_);
}

=requires _build_dsl_keywords

Requires _build_dsl_keywords, as a proxy for being used in a class that consumes
DSL::Tiny::Role.

=cut

requires qw(build_dsl_keywords);

1;


lib/DSL/Tiny/Role.pm
## no critic (RequireUseStrict RequireUseWarnings)
package DSL::Tiny::Role;
## critic
# ABSTRACT: A simple yet powerful DSL builder.

=head1 SYNOPSIS

    # In e.g. MooseDSL.pm, describe a simple DSL.
    package MooseDSL;

    use Moose;  # or use Moo;

    with qw(DSL::Tiny::Role);

    sub build_dsl_keywords {
        return [
            # keywords will be run through curry_method
            qw(argulator return_self clear_call_log),
        ];
    }

    has call_log => (
        clearer => 'clear_call_log',
        default => sub { [] },
        is      => 'rw',
        lazy    => 1
    );

    sub argulator {
        my $self = shift;
        push @{ $self->call_log }, join "::", @_;
    }

    sub return_self { return $_[0] }

    1;

    ################################################################

    # and then in another file you can use that DSL

    use Test::More;
    use Test::Deep;

    use MooseDSL qw( -install_dsl );

    # peek under the covers, get the instance
    my $dsl = return_self;
    isa_ok( $dsl, 'MooseDSL' );

    # test argument handling, single scalar
    argulator("a scalar");
    cmp_deeply( $dsl->call_log, ['a scalar'], 'scalar arg works' );
    clear_call_log;

    # test argument handling, list of args
    argulator(qw(a list of things));
    cmp_deeply( $dsl->call_log, ['a::list::of::things'], 'list arg works' );
    clear_call_log;

    done_testing;

=head1 DESCRIPTION

I<This is an initial release.  It's all subject to rethinking.  Comments
welcome.>

    every time a language advertises "we make writing dsls easy!" i
    read "i'm going to have to learn a new language for every project"

    Jesse Luehrs (@doyster) 3/8/13, 12:11 PM

Domain-specific languages (DSL's) aid in the efficient expression of
configurations, problems and solutions within a particular domain.  While some
DSL's are built from the ground up with custom lexers, parsers,
etc... (e.g. the Unix build tool "make"), other "internal DSL's" (L<Werner
Schuster|http://www.infoq.com/news/2007/06/dsl-or-not>) are distilled from
existing languages and "speak the language of their domain with an accent"
(L<Piers Cawley|http://www.bofh.org.uk/2007/05/19/domain-agnostic-languages>).

A variety of Perl tools and libraries sport domain specific langagues,
e.g. L<Dancer>, L<Module-CPANfile> and L<Module-Install> and the number of
re-implementations of the underlying plumbing is almost exactly equal to the
number of such modules.  These implementations usually devolve into dirty
tricks (e.g. explicit package stash manipulations) and re-invention of several
wheels.

L<DSL::Tiny> packages the common functionality required to implement an
internal DSL, building on powerful foundations (L<Sub::Exporter>) and effective
techniques (L<Moose> and L<Moo> roles) to allow developers to focus on their
domain-specific issues.  It builds on a flexible mechanism for exporting a set
of subroutines into a package; provides a consistent framework for subroutine
currying; and automates the construction of instances, their association with
DSL fragments and the evaluation of those fragments.

In other words, when I needed to build an internal DSL for a project, I was
shocked at how often the basic brushstrokes had been repeated and how often
these implementations dug down and peeked underneath Perl's stashes.  These
modules are my attempt to provide a reusable solution to the problem via
existing high-leverage tools.

=cut

use Moo::Role;

use Sub::Exporter -setup => { groups => { install_dsl => \&_dsl_build, } };

use Data::OptList;
use MooX::Types::MooseLike::Base qw(ArrayRef);
use Params::Util qw(_ARRAYLIKE);
use Sub::Exporter::Util qw(curry_method);

=method import

An import routine generated by Sub::Exporter.

When invoked as a class method (usually via C<use>) with a C<-install_dsl>
argument it will construct a new instance then generate and install a set of
subroutines using the information provided in the instance's C<dsl_keywords>
attribute.

TODO.

=method install_dsl

A synonym for the Sub::Exporter generated import method.  Sounds better when
one uses it to install into an instance.

=cut

BEGIN { *install_dsl = \&import; }

=attr dsl_keywords

Returns an arrayref of dsl keyword info.

It is lazy.  Classes which consume the role are required to supply a builder
named C<_build_dsl_keywords>.

In its canonical form the contents of the array reference are a series of array
references containing keyword_name => { option_hash } pairs, e.g.

  [ [ keyword1 => { as => &generator('method1') } ],
    [ keyword2 => { before => &generator ]
  ]

Generators are as described in the L<Sub::Exporter> documentation.

However, as the contents of this array reference are processed with
Data::OptList there is a great deal of flexibility, e.g.

  [ qw( m1 m2 ), k4 => { as => &generator('some_method' } ]

is equivalent to:

  [ m1 => undef, m2 => undef, k4 => { as => generator('some_method') } ]

Options are optional.  In particular, if no C<as> generator is provided then
the keyword name is presumed to also be the name of a method in the class and
C<Sub::Exporter::Utils::curry_method> will be applied to that method to
generate the coderef for that keyword.  The makes the above equivalent to:

  [ m1 => { as => generator('m1') }, m2 => { as => generator('m2') },
    k4 => { as => generator('some_method') }
  ]

In its simplest form, the keyword arrayref contains a list of method names
relative to class which consumes this role.

  [ qw( m1 m2 ) ]

Supported options include:

=over 4

=item as

=item before

=item after

=back


=cut

has dsl_keywords => (
    is      => 'rw',
    isa     => ArrayRef,
    lazy    => 1,
    builder => 'build_dsl_keywords',
);

=requires build_dsl_keywords

A subroutine, used as the Moo{,se} builder for the L</dsl_keywords> attribute.
It returns an array reference containing information about the methods and
subroutines that implement the keywords in the DSL.

=cut

=method _dsl_build

Private-ish.  Do you really want to be here?

C<_dsl_build> build's up the set of keywords that L<Sub::Exporter> will
install.

It returns a hashref whose keys are names of keywords and whose values are
coderefs implementing the respective behavior.

It can be invoked on a class (a.k.a. as a class method), usually by C<use>.  If
so, a new instance of the class will be constructed and the various keywords
are curried with respect to that instance.

It can be invoked on a class instance, e.g. via an explicit invocation of
L<import> on an instance.  If so, then that instance is used when constructing
the keywords.

=cut

sub _dsl_build {
    my ( $invocant, $group, $arg, $col ) = @_;

    # if not already an instance, create one.
    my $instance = ref $invocant ? $invocant : $invocant->new();

    # fluff up the keyword specification
    my $keywords = Data::OptList::mkopt_hash( $instance->dsl_keywords,
        { moniker => 'keyword list' }, ['HASH'], );

    my %dsl = map { $_ => $instance->_compile_keyword( $_, $keywords->{$_} ) }
        keys $keywords;

    return \%dsl;
}

=method _compile_keyword

Private, go away.

Generate a sub that implements the keyword, taking care of before's and afters.

=cut

sub _compile_keyword {
    my ( $self, $keyword, $args ) = @_;

    # generate code for keyword
    my $code_generator = $args->{as} || curry_method($keyword);
    my $code = $code_generator->( $self, $keyword );

    # generate before code, if any
    # make sure before is an array ref
    # call each generator (if any), save resulting coderefs
    my $before = $args->{before};
    $before = [$before] unless _ARRAYLIKE($before);
    my @before_code = map { $_->($self) } grep { defined $_ } @{$before};

    # generate after code, if any
    my $after = $args->{after};
    $after = [$after] unless _ARRAYLIKE($after);
    my @after_code = map { $_->($self) } grep { defined $_ } @{$after};

    if ( @before_code or @after_code ) {
        my $new_code = sub {
            my @rval;

            $_->(@_) for @before_code;

            # Cribbed from $Class::MOP::Method::Wrapped::_build_wrapped_method
            # not sure that it doesn't have more parens then necessary, but
            # if it works for them...
            (   ( defined wantarray )
                ? (   (wantarray)
                    ? ( @rval = $code->(@_) )
                    : ( $rval[0] = $code->(@_) )
                    )
                : $code->(@_)
            );

            $_->(@_) for @after_code;

            return unless defined wantarray;
            return wantarray ? @rval : $rval[0];
        };
        return $new_code;
    }

    return $code;
}

1;
